### ISSUE TITLE
iRODS search times out

### ISSUE THREAD

**Issue by chStaiger** [UtrechtUniversity/iBridges#37]
When starting a search which takes very long, the connection to the server gets lost. I also verified that this happens when you start the search with an irods.session.iRODSSession instead of the ibridges wrapped session.

```
from ibridges.search import search
from ibridges import Session
import os

session = Session(irods_env_path=os.path.expanduser("~/.irods/irods_environment.json"))
Auth without password

result = search(session, checksum="%")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File /opt/homebrew/lib/python3.10/site-packages/irods/message/__init__.py:185, in _recv_message_in_len(sock, size)
    184 try:
--> 185     buf = sock.recv(size_left, socket.MSG_WAITALL)
    186 except (AttributeError, ValueError):

File /opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py:1289, in SSLSocket.recv(self, buflen, flags)
   1288 if flags != 0:
-> 1289     raise ValueError(
   1290         "non-zero flags not allowed in calls to recv() on %s" %
   1291         self.__class__)
   1292 return self.read(buflen)

ValueError: non-zero flags not allowed in calls to recv() on <class 'ssl.SSLSocket'>

During handling of the above exception, another exception occurred:

TimeoutError                              Traceback (most recent call last)
File /opt/homebrew/lib/python3.10/site-packages/irods/connection.py:114, in Connection.recv(self, into_buffer, return_message, acceptable_errors)
    113 if into_buffer is None:
--> 114     msg = iRODSMessage.recv(self.socket)
    115 else:

File /opt/homebrew/lib/python3.10/site-packages/irods/message/__init__.py:275, in iRODSMessage.recv(sock)
    272 @staticmethod
    273 def recv(sock):
    274     # rsp_header_size = sock.recv(4, socket.MSG_WAITALL)
--> 275     rsp_header_size = _recv_message_in_len(sock, 4)
    276     rsp_header_size = struct.unpack(">i", rsp_header_size)[0]

File /opt/homebrew/lib/python3.10/site-packages/irods/message/__init__.py:187, in _recv_message_in_len(sock, size)
    186 except (AttributeError, ValueError):
--> 187     buf = sock.recv(size_left)
    188 except OSError as e:
    189     #skip only Windows error 10045

File /opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py:1292, in SSLSocket.recv(self, buflen, flags)
   1289         raise ValueError(
   1290             "non-zero flags not allowed in calls to recv() on %s" %
   1291             self.__class__)
-> 1292     return self.read(buflen)
   1293 else:

File /opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py:1165, in SSLSocket.read(self, len, buffer)
   1164     else:
-> 1165         return self._sslobj.read(len)
   1166 except SSLError as x:

TimeoutError: The read operation timed out

During handling of the above exception, another exception occurred:

NetworkException                          Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 result = search(session, checksum="%")

File ~/git-repos/iBridges/ibridges/search.py:57, in search(session, path, checksum, key_vals)
     55     data_query = data_query.filter(kw.LIKE(kw.DATA_CHECKSUM, checksum))
     56 # gather results
---> 57 results = [{key.icat_key: val} for res in data_query.get_results()
     58                                for key, val in res.items()]
     59 if checksum is None:
     60     results.extend([{key.icat_key: val} for res in coll_query.get_results()
     61                                         for key, val in res.items()])

File ~/git-repos/iBridges/ibridges/search.py:57, in <listcomp>(.0)
     55     data_query = data_query.filter(kw.LIKE(kw.DATA_CHECKSUM, checksum))
     56 # gather results
---> 57 results = [{key.icat_key: val} for res in data_query.get_results()
     58                                for key, val in res.items()]
     59 if checksum is None:
     60     results.extend([{key.icat_key: val} for res in coll_query.get_results()
     61                                         for key, val in res.items()])

File /opt/homebrew/lib/python3.10/site-packages/irods/query.py:212, in Query.get_results(self)
    211 def get_results(self):
--> 212     for result_set in self.get_batches():
    213         for result in result_set:
    214             yield result

File /opt/homebrew/lib/python3.10/site-packages/irods/query.py:195, in Query.get_batches(self)
    194 def get_batches(self):
--> 195     result_set = self.execute()
    197     try:
    198         yield result_set

File /opt/homebrew/lib/python3.10/site-packages/irods/query.py:174, in Query.execute(self)
    172 conn.send(message)
    173 try:
--> 174     result_message = conn.recv()
    175     results = result_message.get_main_message(GenQueryResponse)
    176     result_set = ResultSet(results)

File /opt/homebrew/lib/python3.10/site-packages/irods/connection.py:125, in Connection.recv(self, into_buffer, return_message, acceptable_errors)
    123     logger.error("Could not receive server response")
    124     self.release(True)
--> 125     raise NetworkException("Could not receive server response")
    126 if isinstance(return_message,list): return_message[:] = [msg]
    127 if msg.int_info < 0:

NetworkException: Could not receive server response
```

**Comment by a-tsygan** [comment]
We had a similar issue at The University of Groningen and actually spent some time on it. The problem might be a combination of problems. In our case there were two major parameters to setup:
1. Setup connection_timeout at the irods session level: 
        irodssession.connection_timeout = 10000
2. tcp_keep_alive server setting. It needs to be configured on the iCAT server and the client side so that tcp connection is not cut out.

**Comment by chStaiger** [comment]
@a-tsygan  Thank you!!! I added your suggestion to the session code. Hwoever, since on the client machine you also have to set these parameters, it will be very difficult to make sure that all users can benefit from it.

**Comment by chStaiger** [comment]
Should be fixed by #204
